### ISSUE TITLE
StaxRip MP4Box Crash on AAC 7.1 MKV File Open

### ISSUE THREAD

**Issue by anthony75smith75** [staxrip/staxrip#489]
I get StaxRip crash on opening MKV files with AAC 7.1 audio. ALWAYS occurs on ALL MKV files with aac 7.1 audio. MKV files with aac 5.1 audio open without issue. If I leave video and audio untouched and convert the MKV to an MP4 container the file opens fine. The culprit appears to be MP4Box. It reads the 7.1 aac in MKV containers as 5 channels? Then the crash occurs. Please find attached the error log and jpg showing error.

Thanks for your help and dedication on this issue.

------------------------- System Environment -------------------------

StaxRip    : 2.1.7.6
Windows    : Windows 10 Pro 2004
Language   : English (Australia)
CPU        : Intel(R) Core(TM) i5-4590 CPU @ 3.30GHz
GPU        : Intel(R) HD Graphics 4600
Resolution : 2240 x 1260
DPI        : 158
Code Page  : 1252

----------------------- Media Info Source File -----------------------

C:\Encode\4K HEVC 7.1 AAC.mkv

General
Complete name           : C:\Encode\4K HEVC 7.1 AAC.mkv
Format                  : Matroska
Format version          : Version 4
File size               : 5.50 GiB
Duration                : 1 h 58 min
Overall bit rate        : 6 628 kb/s
Encoded date            : UTC 2021-02-04 09:33:02
Writing application     : mkvmerge v52.0.0 ('Secret For The Mad') 64-bit
Writing library         : libebml v1.4.1 + libmatroska v1.6.2
Cover                   : Yes
Attachments             : cover.jpg

Video
ID                      : 1
Format                  : HEVC
Format/Info             : High Efficiency Video Coding
Format profile          : Main 10@L5@Main
Codec ID                : V_MPEGH/ISO/HEVC
Duration                : 1 h 58 min
Bit rate                : 6 002 kb/s
Width                   : 3 840 pixels
Height                  : 1 616 pixels
Display aspect ratio    : 2.40:1
Frame rate mode         : Constant
Frame rate              : 23.976 (24000/1001) FPS
Color space             : YUV
Chroma subsampling      : 4:2:0
Bit depth               : 10 bits
Bits/(Pixel*Frame)      : 0.040
Stream size             : 4.98 GiB (91%)
Writing library         : x265 3.2+9-971180b100f8:[Windows][GCC 9.2.0][64 bit] 10bit
Default                 : Yes
Forced                  : No

Audio
ID                      : 2
Format                  : AAC LC
Format/Info             : Advanced Audio Codec Low Complexity
Codec ID                : A_AAC-2
Duration                : 1 h 58 min
Bit rate                : 609 kb/s
Channel(s)              : 8 channels
Channel layout          : C L R Ls Rs Lw Rw LFE
Sampling rate           : 48.0 kHz
Frame rate              : 46.875 FPS (1024 SPF)
Compression mode        : Lossy
Delay relative to video : 9 ms
Stream size             : 518 MiB (9%)
Title                   : Surround 7.1
Language                : English
Default                 : No
Forced                  : No

Text #1
ID                      : 3
Format                  : PGS
Muxing mode             : zlib
Codec ID                : S_HDMV/PGS
Codec ID/Info           : Picture based subtitle format used on BDs/HD-DVDs
Duration                : 1 h 51 min
Bit rate                : 40.6 kb/s
Count of elements       : 2674
Stream size             : 32.4 MiB (1%)
Language                : English
Default                 : No
Forced                  : No

Text #2
ID                      : 4
Format                  : VobSub
Muxing mode             : zlib
Codec ID                : S_VOBSUB
Codec ID/Info           : Picture based subtitle format used on DVDs
Duration                : 1 h 54 min
Bit rate                : 4 776 b/s
Count of elements       : 917
Stream size             : 3.92 MiB (0%)
Language                : Arabic
Default                 : No
Forced                  : No

Menu
00                      : 00:00.000                             : en:Chapter 01
00                      : 03:08.938                             : en:Chapter 02
00                      : 07:02.171                             : en:Chapter 03
00                      : 13:36.023                             : en:Chapter 04
00                      : 16:43.168                             : en:Chapter 05
00                      : 27:29.939                             : en:Chapter 06
00                      : 36:06.747                             : en:Chapter 07
00                      : 45:25.389                             : en:Chapter 08
00                      : 51:33.423                             : en:Chapter 09
00                      : 58:01.769                             : en:Chapter 10
01                      : 02:57.648                             : en:Chapter 11
01                      : 12:17.666                             : en:Chapter 12
01                      : 17:38.654                             : en:Chapter 13
01                      : 22:44.542                             : en:Chapter 14
01                      : 28:29.304                             : en:Chapter 15
01                      : 37:17.081                             : en:Chapter 16
01                      : 48:14.279                             : en:Chapter 17
01                      : 55:00.727                             : en:Chapter 18

------------------------------ Demux MKV ------------------------------

mkvextract 52

"C:\Encode\StaxRip Beta\Apps\Support\MKVToolNix\mkvextract.exe" "C:\Encode\4K HEVC 7.1 AAC.mkv" tracks 2:"C:\Encode\4K HEVC 7.1 AAC_temp\ID1 English.sup" 3:"C:\Encode\4K HEVC 7.1 AAC_temp\ID2 Arabic.idx" 1:"C:\Encode\4K HEVC 7.1 AAC_temp\ID1 9ms English {Surround 7.1}.aac" --ui-language en

Extracting track 1 with the CodecID 'A_AAC' to the file 'C:\Encode\4K HEVC 7.1 AAC_temp\ID1 9ms English {Surround 7.1}.aac'. Container format: raw AAC file with ADTS headers
Extracting track 2 with the CodecID 'S_HDMV/PGS' to the file 'C:\Encode\4K HEVC 7.1 AAC_temp\ID1 English.sup'. Container format: SUP
Extracting track 3 with the CodecID 'S_VOBSUB' to the file 'C:\Encode\4K HEVC 7.1 AAC_temp\ID2 Arabic.sub'. Container format: VobSubs
Writing the VobSub index file 'C:\Encode\4K HEVC 7.1 AAC_temp\ID2 Arabic.idx'.

Start:    7:42:20 PM
End:      7:42:23 PM
Duration: 00:00:03

General
Complete name    : C:\Encode\4K HEVC 7.1 AAC_temp\ID1 9ms English {Surround 7.1}.aac
Format           : ADTS
Format/Info      : Audio Data Transport Stream
File size        : 520 MiB

Audio
Format           : AAC LC
Format/Info      : Advanced Audio Codec Low Complexity
Format version   : Version 4
Codec ID         : 2
Sampling rate    : 48.0 kHz
Frame rate       : 46.875 FPS (1024 SPF)
Compression mode : Lossy
Stream size      : 520 MiB (100%)

------------------------ Error Mux AAC to M4A ------------------------

Mux AAC to M4A returned error exit code: -1073741819 (0xC0000005)

--------------------------- Mux AAC to M4A ---------------------------

MP4Box 1.1.0-rev447-g8c190b551-gcc10.2.0 Patman

"C:\Encode\StaxRip Beta\Apps\Support\MP4Box\MP4Box.exe" -add "C:\Encode\4K HEVC 7.1 AAC_temp\ID1 9ms English {Surround 7.1}.aac:name= " -new "C:\Encode\4K HEVC 7.1 AAC_temp\ID1 9ms English {Surround 7.1}.m4a"

Track Importing AAC  - SampleRate 7350 Num Channels 5

--------------------------- Mux AAC to M4A ---------------------------

MP4Box 1.1.0-rev447-g8c190b551-gcc10.2.0 Patman

"C:\Encode\StaxRip Beta\Apps\Support\MP4Box\MP4Box.exe" -add "C:\Encode\4K HEVC 7.1 AAC_temp\ID1 9ms English {Surround 7.1}.aac:name= " -new "C:\Encode\4K HEVC 7.1 AAC_temp\ID1 9ms English {Surround 7.1}.m4a"

Track Importing AAC  - SampleRate 7350 Num Channels 5

Start:    7:42:23 PM
End:      7:42:27 PM
Duration: 00:00:04

------------------------------ Exception ------------------------------

StaxRip.ErrorAbortException: Mux AAC to M4A returned error exit code: -1073741819 (0xC0000005)

--------------------------- Mux AAC to M4A ---------------------------

MP4Box 1.1.0-rev447-g8c190b551-gcc10.2.0 Patman

"C:\Encode\StaxRip Beta\Apps\Support\MP4Box\MP4Box.exe" -add "C:\Encode\4K HEVC 7.1 AAC_temp\ID1 9ms English {Surround 7.1}.aac:name= " -new "C:\Encode\4K HEVC 7.1 AAC_temp\ID1 9ms English {Surround 7.1}.m4a"

Track Importing AAC  - SampleRate 7350 Num Channels 5

   at StaxRip.Proc.Start() in D:\Projekte\VB\staxrip\General\Proc.vb:line 374
   at StaxRip.mkvDemuxer.Demux(String sourcefile, IEnumerable`1 audioStreams, IEnumerable`1 subtitles, AudioProfile ap, Project proj, Boolean onlyEnabled, Boolean videoDemuxing, Boolean overrideExisting, String title, Boolean useStreamName) in D:\Projekte\VB\staxrip\General\Demux.vb:line 935
   at StaxRip.mkvDemuxer.Run(Project proj) in D:\Projekte\VB\staxrip\General\Demux.vb:line 721
   at StaxRip.MainForm.Demux() in D:\Projekte\VB\staxrip\Forms\MainForm.vb:line 3232
   at StaxRip.MainForm.OpenVideoSourceFiles(IEnumerable`1 files, Boolean isEncoding) in D:\Projekte\VB\staxrip\Forms\MainForm.vb:line 2142

**Comment by stax76** [comment]
Tool issue, not much we can do, but you can try:

Options > Audio > Demux Audio > None

@jeanlf
@aureliendavid

**Comment by jeanlf** [comment]
could you please open a [gpac issue](https://github.com/gpac/gpac/issues) and upload a faulty AAC file ?

**Comment by stax76** [comment]
Looking at the log file an issue can be seen, after demuxing with mkvextract the media info does not show the channel count of the aac file.

I tried to reproduce it, here the channel count of the demuxed aac file is shown as 8, but if I convert from aac to m4a with mp4box, media info shows only 2 channels in the m4a file, so it looks like we have two independent issues.

Here is my test:

https://www.dropbox.com/s/d8ye38o9fenubda/StaxRip%20Temp.7z?dl=0

This includes both a video file and a log file containing the command lines and media info.

@anthony75smith75 

Please upload a sample.

**Comment by Dendraspis** [comment]
@anthony75smith75 
Do you have created the source file `C:\Encode\4K HEVC 7.1 AAC.mkv` yourself? Can you verify, that it is really a ***7.1 AAC-LC*** file?

**Comment by Dendraspis** [comment]
@stax76 , @jeanlf 
I can confirm, that this seems to be **_neither_** a StaxRip **_nor_** a MP4Box issue!

**Comment by Dendraspis** [comment]
Okay, got it!  🕺 

***mkvextract*** (the tool StaxRip uses for this task) is unable to extract a stream from the file, that ***mkvmerge*** created.  🥇 
Maybe there is a reason for that, but ***WTF*** !!  😡 

The extracted AAC file is broken and comes with wrong values and headers. Interestingly ***ffmpeg*** suffers with ***mkvextract*** and creates an identical (broken) file. 👍 

###  🙄  Workaround 

Extracting the stream into another container is possible without destroying the data, so ***mkvextract*** and ***ffmpeg*** are able to extract into an M4A file that works. Sadly it's not possible to get the AAC out of this M4A file. At least it's the right container for an audio file.

**_Does anybody know other tools, that can demux streams from MKV containers?_**

------------

@stax76 
For StaxRip it should be okay just changing the extension for de- and remuxes.  🤔

**Comment by jeanlf** [comment]
That's my conclusion as well, ffmpeg and mkvextract dumps the aac ADTS using channel config 0 instead of 7 (for 7.1 config). I'm currently patching GPAC to deal with 7.1 signaling in ADTS.

**Comment by stax76** [comment]
And what about the issue in my test, something is wrong there too.

----------------------- Media Info Source File -----------------------

D:\Work\StaxRip Temp\aac-7.1.mkv

General
Complete name                     : D:\Work\StaxRip Temp\aac-7.1.mkv
Format                            : Matroska
Format version                    : Version 4
File size                         : 1.37 MiB
Duration                          : 10 s 729 ms
Overall bit rate                  : 1 070 kb/s
Encoded date                      : UTC 2021-02-04 12:32:22
Writing application               : mkvmerge v52.0.0 ('Secret For The Mad') 64-bit
Writing library                   : libebml v1.4.1 + libmatroska v1.6.2

Video
ID                                : 1
Format                            : HEVC
Format/Info                       : High Efficiency Video Coding
Format profile                    : Main 10@L3@Main
HDR format                        : SMPTE ST 2086, HDR10 compatible
Codec ID                          : V_MPEGH/ISO/HEVC
Duration                          : 10 s 719 ms
Bit rate                          : 638 kb/s
Width                             : 960 pixels
Height                            : 400 pixels
Display aspect ratio              : 2.40:1
Frame rate mode                   : Constant
Frame rate                        : 23.976 (24000/1001) FPS
Color space                       : YUV
Chroma subsampling                : 4:2:0 (Type 2)
Bit depth                         : 10 bits
Bits/(Pixel*Frame)                : 0.069
Stream size                       : 835 KiB (60%)
Writing library                   : x265 3.4+70-90fa80a85:[DJATOM's Mod][Windows][GCC 10.2.1][64 bit] 10bit
Default                           : Yes
Forced                            : No
Color range                       : Limited
Color primaries                   : BT.2020
Transfer characteristics          : PQ
Matrix coefficients               : BT.2020 non-constant
Mastering display color primaries : Display P3
Mastering display luminance       : min: 0.0000 cd/m2, max: 1100 cd/m2

Audio
ID                                : 2
Format                            : AAC LC
Format/Info                       : Advanced Audio Codec Low Complexity
Codec ID                          : A_AAC-2
Duration                          : 10 s 709 ms
Bit rate                          : 425 kb/s
**Channel(s)                        : 8 channels**
Channel layout                    : C L R Ls Rs Lb Rb LFE
Sampling rate                     : 48.0 kHz
Frame rate                        : 46.875 FPS (1024 SPF)
Compression mode                  : Lossy
Delay relative to video           : 20 ms
Stream size                       : 555 KiB (40%)
Language                          : English
Default                           : No
Forced                            : No

------------------------------ Demux MKV ------------------------------

mkvextract 52

D:\Projekte\VB\staxrip\bin\Apps\Support\MKVToolNix\mkvextract.exe "D:\Work\StaxRip Temp\aac-7.1.mkv" tracks 1:"D:\Work\StaxRip Temp\aac-7.1_temp\ID1 20ms English.aac" --ui-language en

Extracting track 1 with the CodecID 'A_AAC' to the file 'D:\Work\StaxRip Temp\aac-7.1_temp\ID1 20ms English.aac'. Container format: raw AAC file with ADTS headers

Start:    13:33:24
End:      13:33:24
Duration: 00:00:00

General
Complete name         : D:\Work\StaxRip Temp\aac-7.1_temp\ID1 20ms English.aac
Format                : ADTS
Format/Info           : Audio Data Transport Stream
File size             : 559 KiB
Overall bit rate mode : Variable

Audio
Format                : AAC LC
Format/Info           : Advanced Audio Codec Low Complexity
Format version        : Version 4
Codec ID              : 2
Bit rate mode         : Variable
**Channel(s)            : 8 channels**
Channel layout        : C L R Ls Rs Lb Rb LFE
Sampling rate         : 48.0 kHz
Frame rate            : 46.875 FPS (1024 SPF)
Compression mode      : Lossy
Stream size           : 559 KiB (100%)

--------------------------- Mux AAC to M4A ---------------------------

MP4Box 1.1.0-rev447-g8c190b551-gcc10.2.0 Patman

D:\Projekte\VB\staxrip\bin\Apps\Support\MP4Box\MP4Box.exe -add "D:\Work\StaxRip Temp\aac-7.1_temp\ID1 20ms English.aac:name= " -new "D:\Work\StaxRip Temp\aac-7.1_temp\ID1 20ms English.m4a"

Track Importing AAC  - SampleRate 32000 Num Channels 2
[ADTSDmx] truncated ADTS frame!
Setting up iTunes/iPod file...
Saving D:\Work\StaxRip Temp\aac-7.1_temp\ID1 20ms English.m4a: 0.500 secs Interleaving

Start:    13:33:24
End:      13:33:25
Duration: 00:00:01


General
Complete name         : D:\Work\StaxRip Temp\aac-7.1_temp\ID1 20ms English.m4a
Format                : MPEG-4
Format profile        : Apple audio with iTunes info
Codec ID              : M4A  (isom/M4A /mp42)
File size             : 12.5 KiB
Duration              : 42 ms
Overall bit rate mode : Constant
Overall bit rate      : 2 433 kb/s
Encoded date          : UTC 2021-02-04 12:33:25
Tagged date           : UTC 2021-02-04 12:33:25

Audio
ID                    : 1
Format                : AAC LTP
Format/Info           : Advanced Audio Codec
Codec ID              : mp4a-40-4
Duration              : 42 ms
Duration_FirstFrame   : 21 ms
Bit rate mode         : Constant
Bit rate              : 2 249 kb/s
**Channel(s)            : 2 channels**
Channel layout        : L R
Sampling rate         : 32.0 kHz
Frame rate            : 31.250 FPS (1024 SPF)
Compression mode      : Lossy
Stream size           : 11.7 KiB (94%)
Title                 : 
Encoded date          : UTC 2021-02-04 12:33:25
Tagged date           : UTC 2021-02-04 12:33:25

**Comment by jeanlf** [comment]
same issue, comes from buggy extracted ADTS

**Comment by stax76** [comment]
Would that be a mkvextract or mkvmerge bug? In that case it should be reported.

**Comment by Dendraspis** [comment]
Not really sure if *mkvmerge* also messes something up, but it is already reported and meanwhile there is also a pull request, but Markus will *"look into it further. No ETA, though."*

**Comment by JJKylee** [comment]
Well, this whole discussion reminds me of [this](https://gitlab.com/mbunkus/mkvtoolnix/-/wikis/Appending-&-splitting-FLAC-audio-tracks-not-supported).
So `mkvextract` cannot extract **ADTS** aac to the _aac_ container properly, and **flac** is also given up...? 😖😔😞😩

**Comment by mbunkus** [comment]
> Markus will "look into it further. No ETA, though."

Moritz, not Markus. Funny thing, that, I've been wrongfully called Markus by almost everyone I know one time or another.

Anyway.

mkvmerge is doing fine. mkvextract simply doesn't support it. The problem with ADTS AAC files and higher channel counts is that ADTS AAC headers only support certain channel configurations (it's only a three-bit field). Therefore there's something called a `program_config_element` which might be present right at the start of the AAC bitstream (meaning directly behind the first ADTS AAC header). mkvmerge can parse such a structure, but mkvextract cannot create it.

The proposed PR likely works for 7.1, but not for anything else (e.g. 7.0 channels or 8.0 or more than 8 channels in total).

When in doubt, don't use mkvextract for extracting AAC.

**Comment by Dendraspis** [comment]
@mbunkus 
Entschuldige.

Really frustrating such issues. 😞 
Regarding the PR a working 7.1 ist better than nothing and 7.1 is far more likely and usual than 7.0 or 8.0 - and has anybody seen 9+ channels in real life?  A support would be really nice to get a chance extracting them.

**Comment by mbunkus** [comment]
> Entschuldige.

No worries, I've gotten so used to it I usually react to someone calling me Markus :grin:

> Regarding the PR a working 7.1 ist better than nothing and 7.1 is far more likely and usual than 7.0 or 8.0 - and has anybody seen 9+ channels in real life?

I have actually merged the PR righter after I posted my comment earlier.

And it's true that 7.1 is used far more often. I do have 7.0 files, though, and as I said, those cannot be extracted by mkvextract as they require a `program_config_element`.

I don't have any test files with more than 8 channels at the moment.

**Comment by anthony75smith75** [comment]
If anyone needs a test file:

https://drive.google.com/file/d/1G-2cDppaSHlJKQLcQVOY9M2va-TlpUUI/view?usp=sharing

**Comment by JJKylee** [comment]
Thanks.

BTW, it looks like an [upscaled fake 4K Star Trek video](https://4kmedia.org/real-or-fake-4k/5/). 🤨
No offense. Please don't get me wrong. 😅

**Comment by anthony75smith75** [comment]
Good quality picture (film source) with the original re-masted 7.1 sound causing all the issues as presented here.

https://www.blu-ray.com/Star-Trek-IV-The-Voyage-Home/20277/

**Comment by Dendraspis** [comment]
Seems like @Patman86 is out of business - the MediaFire folder is empty. 😟

Where do we get a fresh build from to close this issue?  😕

**Comment by Patman86** [comment]
I'll be uploading updates tonight 😉

**Comment by Patman86** [comment]
At the weekend I'll be working on my mods for x264 and x265, then I'll be involved again 😁😂

**Comment by 44vince44** [comment]
Now that's really good news !! Even though it's much work, I hope it's gonna be smooth and easy !

> I'll be involved again

Oh come on! You've always been !

**Comment by 44vince44** [comment]
@stax76 there is a recent release of MP4BOX by @Patman86 that is not yet updated in Staxrip, can you please update?
Thanks !
so we can close this thread i guess....

**Comment by stax76** [comment]
I've updated it now.

**Comment by Dendraspis** [comment]
...and I've closed it now. 😁

**Comment by JJKylee** [comment]
MKVToolNix 54.0.0 is released today, and it seems that this version addresses this issue.

> mkvextract: AAC: fixed wrong channel mask field in the ADTS headers for 7.1 channel layouts. Fix by Tom Yan. Fixes #2636.

Let's see if this update fixes this issue and other related ones.

**Comment by JJKylee** [comment]
Combined with @Patman86's `Mp4Box-1.1.0-DEV-rev508-g3106831ad-x64-gcc10.2.0.7z`, the new `mkvextract` (v54.0.0) **_does_** resolve this issue.

```
"D:\Utilities\StaxRip\Apps\Support\MP4Box\MP4Box.exe" -add "D:\Work\tmp\tst\4K HEVC 7.1 AAC_temp\ID1 14ms English {Surround 7.1}.aac:name= " -new "D:\Work\tmp\tst\4K HEVC 7.1 AAC_temp\ID1 14ms English {Surround 7.1}.m4a"

Track Importing AAC  - SampleRate 48000 Num Channels 8
Setting up iTunes/iPod file= | (99/100)
Saving D:\Work\tmp\tst\4K HEVC 7.1 AAC_temp\ID1 14ms English {Surround 7.1}.m4a: 0.500 secs Interleaving
```

No errors now! 🎉
And the output `m4a` has the correct number of channels and plays OK.
